apps/cowswap-frontend/src/modules/hooksStore/hooks/useBalancesDiff.ts

```diff
@@ -0,0 +1,124 @@
+    const lastPreHook = preHooks[preHooks.length - 1]
+    return data[lastPreHook?.uuid]?.cumulativeBalancesDiff || EMPTY_BALANCE_DIFF
+  }, [data, preHooks])
+}
+
+// Returns all the ERC20 Balance Diff of the current hook to be passed to the iframe context
+export function useHookBalancesDiff(isPreHook: boolean, hookToEditUid?: string): BalancesDiff {
+  const { account } = useWalletInfo()
+  const { data } = useTenderlyBundleSimulation()
+  const orderParams = useOrderParams()
+  const { preHooks, postHooks } = useHooks()
+  const preHookBalanceDiff = usePreHookBalanceDiff()
+
+  const orderMockBalanceDiff = useMemo(() => {
+    if (!account) return EMPTY_BALANCE_DIFF
+    const balanceDiff: Record<string, string> = {}
+
+    if (orderParams?.buyAmount && orderParams.buyTokenAddress && account)
+      balanceDiff[orderParams.buyTokenAddress.toLowerCase()] = orderParams.buyAmount
+
+    if (orderParams?.sellAmount && orderParams.sellTokenAddress && account)
+      balanceDiff[orderParams.sellTokenAddress.toLowerCase()] = `-${orderParams.sellAmount}`
+
+    return { account: balanceDiff }
+  }, [orderParams, account])
+
+  const firstPostHookBalanceDiff = useMemo(() => {
+    return mergeBalanceDiffs(preHookBalanceDiff, orderMockBalanceDiff)
+  }, [preHookBalanceDiff, orderMockBalanceDiff])
+
+  const postHookBalanceDiff = useMemo(() => {
+    // is adding the first post hook or simulation not available
+    if (!data || !postHooks) return firstPostHookBalanceDiff
+
+    const lastPostHook = postHooks[postHooks.length - 1]
+    return data[lastPostHook?.uuid]?.cumulativeBalancesDiff || firstPostHookBalanceDiff
+  }, [data, postHooks, orderMockBalanceDiff, preHookBalanceDiff])
+
+  const hookToEditBalanceDiff = useMemo(() => {
+    if (!data || !hookToEditUid) return EMPTY_BALANCE_DIFF
+
+    const otherHooks = isPreHook ? preHooks : postHooks
+
+    const hookToEditIndex = otherHooks.findIndex((hook) => hook.uuid === hookToEditUid)
+
+    // is editing first preHook -> return empty state
+    if (!hookToEditIndex && isPreHook) return EMPTY_BALANCE_DIFF
+
+    // is editing first postHook -> return
+    if (!hookToEditIndex && !isPreHook) return firstPostHookBalanceDiff
+
+    // is editing a non first hook, return the latest available hook state
+    const previousHookIndex = hookToEditIndex - 1
+
+    return data[otherHooks[previousHookIndex]?.uuid]?.cumulativeBalancesDiff || EMPTY_BALANCE_DIFF
+  }, [data, hookToEditUid, isPreHook, preHooks, postHooks, firstPostHookBalanceDiff])
+
+  return useMemo(() => {
+    if (hookToEditUid) return hookToEditBalanceDiff
+    if (isPreHook) return preHookBalanceDiff
+    return postHookBalanceDiff
+  }, [data, orderParams, preHooks, postHooks])
+}
+
+const addBigNumberStrings = (a: string, b: string): string => {
+  const bigA = BigNumber.from(a)
+  const bigB = BigNumber.from(b)
+  return bigA.add(bigB).toString()
+}
+
+function mergeBalanceDiffs(first: BalancesDiff, second: BalancesDiff): BalancesDiff {
```

Comment:
Checking the type of the `BalancesDiff` I see is a structure to hold token balances. Having `Diff` in its name feels a bit constraining cause you can use it for example to hold the current balances, etc.

I would drop the `diff` from the name and would make the merging utilities more generic so they live not in the `hookStore` but in a more central place. Or at least can be moved to a utility instead of being in a hook file

Reply:
I think the main difference between this and a token balances is that it can be negative. Another difference between this and the token balances used on the rest of the app is the first key. Another difference is that the outer key of `BalanceDiff` is the `address` which the difference is related to while the `BalanceState` of the rest of the app the outer key is the `chainId` since it is always related to the connected account. For now, I don't see any other place on the app where we would store balances diff from addresses, so I think that there isn't too much gain of moving this to a utility for now.

Reply:
ok, makes sense